Lib/fractions.py

```diff
@@ -21,17 +21,17 @@
 _PyHASH_INF = sys.hash_info.inf
 
 _RATIONAL_FORMAT = re.compile(r"""
-    \A\s*                      # optional whitespace at the start, then
-    (?P<sign>[-+]?)            # an optional sign, then
-    (?=\d|\.\d)                # lookahead for digit or .digit
-    (?P<num>\d*)               # numerator (possibly empty)
-    (?:                        # followed by
-       (?:/(?P<denom>\d+))?    # an optional denominator
-    |                          # or
-       (?:\.(?P<decimal>\d*))? # an optional fractional part
-       (?:E(?P<exp>[-+]?\d+))? # and optional exponent
+    \A\s*                                    # optional whitespace at the start,
+    (?P<sign>[-+]?)                          # an optional sign, then
+    (?=\d|\.\d)                              # lookahead for digit or .digit
+    (?P<num>((\d+_?)*\d|\d*))                # numerator (possibly empty)
```

Comment:
There's potential for pathologically bad backtracking behaviour with this kind of +-within-* or *-within-* regex. We can either find a way to effectively flatten the regex to avoid that,  or revert to your original approach of accepting arbitrary substrings of underscore-or-digit and then making use of the `int` calls to further validate.

Either way, I'd suggest adding something like `Fraction("9"*50 + "_")` as a test case.

Reply:
I would prefer to keep regexp-based approach, if it's possible.  I think, the updated version doesn't suffer from catastrophic backtracking.  At least, your test doesn't hangs.

I've added tests, that might trigger this problem in den/decimal/exp as well.  This makes some tests above - redundant.  Maybe I should remove them?

Reply:
> I think, the updated version doesn't suffer from catastrophic backtracking.

Agreed, I think the new version should be fine.

Sure, removing redundant tests sounds fine to me.

Reply:
Ok, I did.